### PR TITLE
Omit joining initial cut from CutDB with genesis cut

### PR DESCRIPTION
**Pull request by larskuhtz**
We generally trust the chain db. Let's also trust the cut that is persisted in the DB.

### CODE REVIEW

**Comment by larskuhtz** [comment]
Tested and confirmed that the fix works.

File: src/Chainweb/CutDB.hs [hunk 1]
```diff
@@ -407,15 +408,13 @@ startCutDb config logfun headerStore payloadStore cutHashesStore = mask_ $ do
...
-                    hm' <- case _cutDbParamsInitialHeightLimit config of
+                Right hm -> unsafeMkCut v
+                    <$> case _cutDbParamsInitialHeightLimit config of
                         Nothing -> return hm
                         Just h -> limitCutHeaders wbhdb h hm
-                    logg Debug $ "joining intial cut from cut db configuration with cut that was loaded from the database " <> sshow hm'
-                    joinIntoHeavier_
-                        (_webBlockHeaderStoreCas headerStore)
-                        hm'
-                        (_cutMap $ _cutDbParamsInitialCut config)
```
Comment (larskuhtz) [inline src/Chainweb/CutDB.hs]: This was joining the persisted cut with the genesis cut. By joining the cuts, it was validated that the stored cut was actually a cut (correctly braided) and all headers in the history of the cut exist in the chain db.

The algorithm for joining cuts creates a heap the includes all headers in the symmetric difference of the involved cuts. In the case of the join with the genesis cut, this is all headers in the history of the chain, which at the current block height in mainnet requires > 8GB. Nodes with <=8GB would run out of memory and crash.

By removing this step we just trust the stored cut in the DB. Which should be fine, since we trust the db in any case.

There are more efficient (and more complete) ways to validate the integrity of the DB, which can be performed offline. It would also be possible to add support to do this in the background in a chainweb node. But that's beyond the scope of the PR.

File: src/Chainweb/CutDB.hs [hunk 2]
```diff
@@ -407,15 +408,13 @@ startCutDb config logfun headerStore payloadStore cutHashesStore = mask_ $ do
...
+                            -- FIXME: this requires that the pact db is deleted because it
+                            -- interfers with rewind limit. As a fix temporarily disable
+                            -- rewind limit during initialization.
```
Comment (larskuhtz) [inline src/Chainweb/CutDB.hs]: This FIXME is unrelated to the change in this PR. But we discovered the issue while working on this PR. It's not urgent, since the initial cut limit is rarely used.

File: src/Chainweb/CutDB.hs (hunk 1 above)
Comment (slpopejoy) [reply src/Chainweb/CutDB.hs]: The old check would seem not just to be RAM intensive but probably take a while, no?

Also is it truly redundant? Are there any modes where the check would alert of a corrupted node?

File: src/Chainweb/Cut.hs [hunk 3]
```diff
@@ -434,7 +441,7 @@ join_ wdb prioFun a b = do
         db <- getWebBlockHeaderDb wdb cid
         (q' :> h) <- S.fold g q id $ branchDiff_ db x y
         let !h' = q' `seq` h `seq` HM.insert cid h m
-        return $! (h', q')
+        return (h', q')
```
Comment (slpopejoy) [inline src/Chainweb/Cut.hs]: why remove the strictness?

File: src/Chainweb/CutDB.hs (hunk 1 above)
Comment (larskuhtz) [reply src/Chainweb/CutDB.hs]: The check would capture some database corruptions but not all. I see little value in it. In particular, it the failure is just an invalid braiding, it would be corrected as soon as the next remote cut is received. Only, if all nodes would have invalidly braided cuts it would be an issue.

If a block header is missing, the node would either work normally, if the header isn't used or it would fail if it tries to access the header. Depending on the situation the node may or may not recover from it by querying the missing header from a remote node.

Comment (larskuhtz) [reply src/Chainweb/CutDB.hs]: On nodes with sufficient RAM, the test is surprisingly fast -- still it takes some time. The runtime seems to scales super-linear under RAM pressure, probably due to caching issues.

However, DB pruning seems to be the main runtime bottleneck during startup. In particular on nodes with low number of cores (since it's concurrent for chains).

File: src/Chainweb/Cut.hs (hunk 3 above)
Comment (larskuhtz) [reply src/Chainweb/Cut.hs]: It's redundant. `(,)` is a constructor and constructor application is already in weak head normal form, so it's always strict.

Comment (slpopejoy) [reply src/Chainweb/Cut.hs]: but ... strictness theater! 😄

File: src/Chainweb/CutDB.hs (hunk 1 above)
Comment (slpopejoy) [reply src/Chainweb/CutDB.hs]: I asked about time, b/c if nodes were taking this time, we could probably design a memory-efficient iterative version of this check, but it sounds like a different/more exhaustive check (and probably offline) is better.

**Comment by fosskers** [comment]
I'm going to test this on my personal node this morning.
